src/htm/encoders/SimHashDocumentEncoder.README.md

```diff
@@ -0,0 +1,132 @@
+
+SimHash is a Locality-Sensitive Hashing (LSH) algorithm from the world of
+nearest-neighbor document similarity search. It is used by the GoogleBot Web
+Crawler to find near-duplicate web pages.
+
+"Similarity" here refers to bitwise similarity (small hamming distance, high
+overlap), not semantic similarity (encodings for "apple" and "computer" will
+have no relation here). Internally, hamming distances are never considered or
+adjusted -- they're always the result of a kind of dynamic statistical
+distribution.
+
+See Also:
+* Code: `./SimHashDocumentEncoder.hpp`
+* Example: `py/htm/examples/encoders/simhash_document_encoder.py`
+* Tests: `bindings/py/tests/encoders/simhash_document_encoder_test.py`
+
+Learn More:
+* Wikipedia: [Locality-Sensitive Hashing](https://en.wikipedia.org/wiki/Locality-sensitive_hashing)
+* Wikipedia: [SimHash](https://en.wikipedia.org/wiki/SimHash)
+* YouTube: [SimHashing Video](https://www.youtube.com/watch?v=gnraT4N43qo)
+
+
+# Example
+
+> **Note on Hashing:**
+Imagine below that we are using a tiny hashing function which outputs a 9-bit
+long binary string digest. (In reality, we use the SHA3 hashing function, with
+the SHAKE256 XOF to give us a binary string digest output which is the same
+length as our output SDR encoding.)
+
+Let's say you have a document: **["hello", "there", "world"]**.
+
+Tokens:
+
+```
+hello
+there
+world
+```
+
+
+# Parameter `tokenSimilarity` OFF:
+
+### Hash Tokens:
+
+```
+hello => [ 0 1 0 1 0 1 0 1 1 ]
+there => [ 0 0 0 1 0 0 1 0 0 ]
+world => [ 1 0 0 1 0 1 0 1 0 ]
+```
+
+### Prepare Token Hashes for Summation (convert 0 to -1):
+
+```
+hello => [ -1 +1 -1 +1 -1 +1 -1 +1 +1 ]
+there => [ -1 -1 -1 +1 -1 -1 +1 -1 -1 ]
+world => [ +1 -1 -1 +1 -1 +1 -1 +1 -1 ]
+```
+
+### Document Summation:
+
+```
+sums  => [ -1 -1 -3 +3 -3 +1 -1 +1 -1 ]
+```
+
+### Document SimHash (filter >= 0, or choose top N sums for sparse):
+
+```
+sim   => [  0  0  0  1  0  1  0  1  0 ]
+```
+
+
+# Parameter `tokenSimilarity` ON:
+
+### Hash Tokens *AND individual Characters*:
+
+```
+hello => [ 0 1 0 1 0 1 0 1 1 ]
+    h => [ 1 0 1 1 0 1 0 0 1 ]
+    e => [ 0 1 0 1 1 0 0 0 1 ]
+    l => [ 1 1 0 1 1 0 1 0 0 ]
+    l => [ 1 1 0 1 1 0 1 0 0 ]
+    o => [ 1 1 1 0 0 0 0 0 0 ]
+there => [ 0 0 0 1 0 0 1 0 0 ]
+    t => [ 1 1 1 0 1 0 1 0 1 ]
+    h => [ 1 0 1 1 0 1 0 0 1 ]
+    e => [ 0 1 0 1 1 0 0 0 1 ]
+    r => [ 0 0 0 1 1 0 0 0 0 ]
+    e => [ 0 1 0 1 1 0 0 0 1 ]
+world => [ 1 0 0 1 0 1 0 1 0 ]
+    w => [ 0 0 0 0 1 1 1 0 1 ]
+    o => [ 1 1 1 0 0 0 0 0 0 ]
+    r => [ 0 0 1 1 0 0 0 1 1 ]
+    l => [ 1 1 0 1 1 0 1 0 0 ]
+    d => [ 0 1 0 0 1 0 0 0 0 ]
+```
+
+### Prepare Token/Char Hashes for Summation (convert 0 to -1):
+
+```
+hello => [ -1 +1 -1 +1 -1 +1 -1 +1 +1 ]
+    h => [ +1 -1 +1 +1 -1 +1 -1 -1 +1 ]
+    e => [ -1 +1 -1 +1 +1 -1 -1 -1 +1 ]
+    l => [ +1 +1 -1 +1 +1 -1 +1 -1 -1 ]
+    l => [ +1 +1 -1 +1 +1 -1 +1 -1 -1 ]
+    o => [ +1 +1 +1 -1 -1 -1 -1 -1 -1 ]
```

Comment:
OT: to my problem with `tokenSimilarity` (that it can skew the representation towards a letter (as in `o` in `looooooooooove`)). What if we do (any/which of these?): 
- just use a single occurence of each letter (set): `hello -> h, e, l, o`
- do tokenization (to letters) for similarity only internally to the token (word): tokenize to letters -> sum -> apply bounds -> binarize to 0/1 representation. This becomes representation of the token and is used for the document (sums) 
- there should be check that similarity=ON representations do not differ too much from original. So "hello" and {"hello", h, e, l, l, o} should have high overlap in the encodings.
  - I'm afraid that would not be the case. Iff, how to solve? 
    - apply `weight` to letter encodings? [w=0.1]:
```
hello => [ -1 +1 -1 +1 -1 +1 -1 +1 +1 ]
    h => [ +0.1 -0.1 +0.1 +0.1 -0.1 +0.1 -0.1 -0.1 +0.1 ]
```
`w` could be such that: `w=1.0/count(letters in word) -0.2` 

Also, words in white/black list should be compared after similarity? (ie we correct spelling errors and then check for include/exclude lists)?

Reply:
>     * just use a single occurence of each letter (set): `hello -> h, e, l, o`

OK, I'll add some deduplication options for both tokens/letters.

>     * do tokenization (to letters) for similarity only internally to the token (word): tokenize to letters -> sum -> apply bounds -> binarize to 0/1 representation. This becomes representation of the token and is used for the document (sums)

I had it like this first, but the current method ended up with better output results (bit usage, etc).

>     * there should be check that similarity=ON representations do not differ too much from original. So "hello" and {"hello", h, e, l, l, o} should have high overlap in the encodings.

I currently bump the weight of the word up a little extra in relation to the letters for just this purpose.